### PR TITLE
[SPARK-6444][SQL] Check argument types for SQL functions

### PR DESCRIPTION
**Pull request by cloud-fan**
it's based on https://github.com/apache/spark/pull/4685, but didn't take in the optimization of avoding lazy val(we can do it in a seperate PR)

The code may looks complicated as I try to abstract some common patterns to reduce code size:
- abstract the usage of holding a lazy val of Numeric or Ordering
- abstract the constraint of type equals for left and right child
- abstract the type check for numeric, bitwise, ordering, etc.

[Review on Reviewable](https://reviewable.io/reviews/apache/spark/6152)

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @liancheng @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -81,7 +81,7 @@ abstract class Expression extends TreeNode[Expression] {
 abstract class BinaryExpression extends Expression with trees.BinaryNode[Expression] {
   self: Product =>
 
-  def symbol: String
+  def symbol: String = sys.error(s"BinaryExpressions must either override eval or evalInternal")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: how is this error message related to symbol?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: my mistake...

**Comment by rxin** [comment]
@cloud-fan can you write up some high level description of the change, i.e. important type hierarchy changes, and introduction of new concepts?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/typeCheckHelpers.scala [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+trait TypeEqualConstraint {
+  self: Expression =>
+
+  protected def typeChecker: DataType => Boolean = _ => true
+
+  def left: Expression
+  def right: Expression
+
+  final override lazy val resolved =
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/typeCheckHelpers.scala]: I was thinking just having a new function for checking types, and don't rely on "resolved", since they are really doing different things (one for whether some child-expression is unresolved, the other for type mismatch)

**Comment by cloud-fan** [comment]
Hi @rxin , seems it's better to work on [SPARK-7562](https://issues.apache.org/jira/browse/SPARK-7562) and this one together. I'll write a design doc today and reopen this PR later.
